{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install --user git+https://github.com/gaarangoa/METk.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "# ------------\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "from metaCH.src.feature_extraction.extract_features import extract_features\n",
    "from metaCH.src.classification.inference import inference\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "example_cfDNA_input=pd.read_csv('example_input.csv')\n",
    "features = extract_features(example_cfDNA_input, 'paper')\n",
    "results = inference(features, 'paper')\n",
    "threshold=0.5\n",
    "results['prediction_labels']=np.where(results['ch_score_meta'] > threshold, 'Blood', 'Tumor')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_Position</th>\n",
       "      <th>End_Position</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Allele_1</th>\n",
       "      <th>Tumor_Allele_2</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>Gene_Name</th>\n",
       "      <th>Reference_Genome</th>\n",
       "      <th>VAF</th>\n",
       "      <th>CANCER_TYPE</th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>ch_score_meta</th>\n",
       "      <th>prediction_labels</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89692905</td>\n",
       "      <td>89692905</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>0.000971</td>\n",
       "      <td>Gastrointestinal Stromal Tumor</td>\n",
       "      <td>CGST62P</td>\n",
       "      <td>0.705218</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89692905</td>\n",
       "      <td>89692905</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>Gastrointestinal Stromal Tumor</td>\n",
       "      <td>CGST62P1</td>\n",
       "      <td>0.723183</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89692926</td>\n",
       "      <td>89692926</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>0.002351</td>\n",
       "      <td>Gastrointestinal Stromal Tumor</td>\n",
       "      <td>CGST45P2</td>\n",
       "      <td>0.671160</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89692926</td>\n",
       "      <td>89692926</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>0.000666</td>\n",
       "      <td>Gastrointestinal Stromal Tumor</td>\n",
       "      <td>CGST45P1</td>\n",
       "      <td>0.787933</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>89717698</td>\n",
       "      <td>89717698</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>SNP</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>PTEN</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>0.001824</td>\n",
       "      <td>Gastrointestinal Stromal Tumor</td>\n",
       "      <td>CGST142P</td>\n",
       "      <td>0.781916</td>\n",
       "      <td>Blood</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start_Position  End_Position Reference_Allele Tumor_Allele_1  \\\n",
       "0      chr10        89692905      89692905                G              A   \n",
       "1      chr10        89692905      89692905                G              A   \n",
       "2      chr10        89692926      89692926                C              A   \n",
       "3      chr10        89692926      89692926                C              A   \n",
       "4      chr10        89717698      89717698                T              A   \n",
       "\n",
       "  Tumor_Allele_2 Variant_Type Hugo_Symbol Gene_Name Reference_Genome  \\\n",
       "0              A          SNP        PTEN      PTEN           GRCh37   \n",
       "1              A          SNP        PTEN      PTEN           GRCh37   \n",
       "2              A          SNP        PTEN      PTEN           GRCh37   \n",
       "3              A          SNP        PTEN      PTEN           GRCh37   \n",
       "4              A          SNP        PTEN      PTEN           GRCh37   \n",
       "\n",
       "        VAF                     CANCER_TYPE Sample_ID  ch_score_meta  \\\n",
       "0  0.000971  Gastrointestinal Stromal Tumor   CGST62P       0.705218   \n",
       "1  0.000200  Gastrointestinal Stromal Tumor  CGST62P1       0.723183   \n",
       "2  0.002351  Gastrointestinal Stromal Tumor  CGST45P2       0.671160   \n",
       "3  0.000666  Gastrointestinal Stromal Tumor  CGST45P1       0.787933   \n",
       "4  0.001824  Gastrointestinal Stromal Tumor  CGST142P       0.781916   \n",
       "\n",
       "  prediction_labels  \n",
       "0             Blood  \n",
       "1             Blood  \n",
       "2             Blood  \n",
       "3             Blood  \n",
       "4             Blood  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "data2",
   "language": "python",
   "name": "data2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
